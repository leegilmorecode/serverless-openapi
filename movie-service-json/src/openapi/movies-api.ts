import { errorResponse, movieSchema } from '../shared/schemas/';

import { createMovieSchema } from '../handlers/create-movie/create-movie.schema';
import { functionNames } from '../shared/function-names/function-names';
import { version } from '../shared/open-api-info/schema-version';

type JSONValue = string | number | boolean | null | JsonObject | JsonArray;
interface JsonObject {
  [key: string]: JSONValue;
}
type JsonArray = JSONValue[];

// redact properties from a dynamic json object
function deletePropertyRecursively(
  obj: JsonObject,
  parentPropertyName: string
): void {
  for (const key in obj) {
    if (typeof obj[key] === 'object') {
      if (key === parentPropertyName) {
        delete obj[key];
      } else {
        deletePropertyRecursively(obj[key] as JsonObject, parentPropertyName);
      }
    }
  }
}

function removeDynamicObjects(
  json: JsonObject,
  parentPropertyName: string
): JsonObject {
  for (const key in json) {
    if (typeof json[key] === 'object') {
      deletePropertyRecursively(json[key] as JsonObject, parentPropertyName);
    }
  }

  return json;
}

// redact properties from a dynamic json schema i.e. the 'x-amazon-apigateway-integration' values
// as we don't want to leak this information to people viewing our openapi spec online
export const getApiJsonRedacted = (accountId: string): Record<string, any> => {
  const json: Record<string, any> = getApiJson(accountId);
  return removeDynamicObjects(json, 'x-amazon-apigateway-integration');
};

// get the raw api json for our integrations
export const getApiJson = (accountId: string): Record<string, any> => {
  return {
    openapi: '3.0.3',
    'x-amazon-apigateway-request-validators': {
      validation: {
        validateRequestBody: true,
        validateRequestParameters: true,
      },
    },
    info: {
      title: 'Movie API Example',
      version,
      description: 'An API for movies',
    },
    paths: {
      [`/${version}/movies`]: {
        get: {
          'x-amazon-apigateway-request-validator': 'validation',
          'x-amazon-apigateway-integration': {
            type: 'aws_proxy',
            httpMethod: 'POST',
            uri: {
              'Fn::Sub': `arn:aws:apigateway:\${AWS::Region}:lambda:path/2015-03-31/functions/\${${functionNames.listMovies}.Arn}/invocations`,
            },
            passthroughBehavior: 'when_no_match',
            credentials: `arn:aws:iam::${accountId}:role/apiRole`,
            responses: {
              default: {
                statusCode: '200',
                responseTemplates: {
                  'application/json': '',
                },
              },
            },
          },
          summary: 'Get all movies',
          responses: {
            '200': {
              description: 'Successful response',
              content: {
                'application/json': {
                  schema: {
                    type: 'array',
                    items: {
                      $ref: '#/components/schemas/Movie',
                    },
                  },
                },
              },
            },
            '400': {
              description: 'Bad Request',
              content: {
                'application/json': {
                  schema: {
                    $ref: '#/components/schemas/ErrorResponse',
                  },
                },
              },
            },
            '404': {
              description: 'Not Found',
              content: {
                'application/json': {
                  schema: {
                    $ref: '#/components/schemas/ErrorResponse',
                  },
                },
              },
            },
          },
        },
        post: {
          'x-amazon-apigateway-request-validator': 'validation',
          'x-amazon-apigateway-integration': {
            type: 'aws_proxy',
            httpMethod: 'POST',
            uri: {
              'Fn::Sub': `arn:aws:apigateway:\${AWS::Region}:lambda:path/2015-03-31/functions/\${${functionNames.createMovie}.Arn}/invocations`,
            },
            passthroughBehavior: 'when_no_match',
            credentials: `arn:aws:iam::${accountId}:role/apiRole`,
            responses: {
              default: {
                statusCode: '200',
                responseTemplates: {
                  'application/json': '',
                },
              },
            },
          },
          summary: 'Create a new movie',
          requestBody: {
            required: true,
            content: {
              'application/json': {
                schema: {
                  $ref: '#/components/schemas/NewMovie',
                },
              },
            },
          },
          responses: {
            '200': {
              description: 'Successful response',
              content: {
                'application/json': {
                  schema: {
                    $ref: '#/components/schemas/Movie',
                  },
                },
              },
            },
            '400': {
              description: 'Bad Request',
              content: {
                'application/json': {
                  schema: {
                    $ref: '#/components/schemas/ErrorResponse',
                  },
                },
              },
            },
            '404': {
              description: 'Not Found',
              content: {
                'application/json': {
                  schema: {
                    $ref: '#/components/schemas/ErrorResponse',
                  },
                },
              },
            },
          },
        },
      },
      [`/${version}/movies/{id}`]: {
        get: {
          'x-amazon-apigateway-request-validator': 'validation',
          'x-amazon-apigateway-integration': {
            type: 'aws_proxy',
            httpMethod: 'POST',
            uri: {
              'Fn::Sub': `arn:aws:apigateway:\${AWS::Region}:lambda:path/2015-03-31/functions/\${${functionNames.getMovieById}.Arn}/invocations`,
            },
            passthroughBehavior: 'when_no_match',
            credentials: `arn:aws:iam::${accountId}:role/apiRole`,
            responses: {
              default: {
                statusCode: '200',
                responseTemplates: {
                  'application/json': '',
                },
              },
            },
          },
          summary: 'Get a movie by id',
          parameters: [
            {
              name: 'id',
              in: 'path',
              description: 'ID of the movie',
              required: true,
              schema: {
                type: 'string',
              },
            },
          ],
          responses: {
            '200': {
              description: 'Successful response',
              content: {
                'application/json': {
                  schema: {
                    $ref: '#/components/schemas/Movie',
                  },
                },
              },
            },
            '400': {
              description: 'Bad Request',
              content: {
                'application/json': {
                  schema: {
                    $ref: '#/components/schemas/ErrorResponse',
                  },
                },
              },
            },
            '404': {
              description: 'Not Found',
              content: {
                'application/json': {
                  schema: {
                    $ref: '#/components/schemas/ErrorResponse',
                  },
                },
              },
            },
          },
        },
      },
    },
    components: {
      schemas: {
        ErrorResponse: errorResponse,
        NewMovie: createMovieSchema,
        Movie: movieSchema,
      },
    },
  };
};
