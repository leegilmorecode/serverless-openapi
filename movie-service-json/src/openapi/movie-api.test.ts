import { getApiJson, getApiJsonRedacted } from './movies-api';

describe('get-api-json', () => {
  it('should return the correct json', () => {
    expect(getApiJson('account-id')).toMatchSnapshot();
  });

  it('should return the redacted version of the json', () => {
    expect(getApiJsonRedacted('account-id')).toMatchSnapshot();
  });
});
