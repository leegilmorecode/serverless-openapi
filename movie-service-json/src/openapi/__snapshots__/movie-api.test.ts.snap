// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get-api-json should return the correct json 1`] = `
{
  "components": {
    "schemas": {
      "ErrorResponse": {
        "properties": {
          "message": {
            "type": "string",
          },
        },
        "required": [
          "message",
        ],
        "type": "object",
      },
      "Movie": {
        "properties": {
          "id": {
            "description": "The movie ID (numeric characters only)",
            "pattern": "^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-4[0-9a-fA-F]{3}-[89aAbB][0-9a-fA-F]{3}-[0-9a-fA-F]{12}$",
            "type": "string",
          },
          "rating": {
            "description": "The rating of the movie",
            "enum": [
              "U",
              "PG",
              "12",
              "15",
              "18",
            ],
            "pattern": "^[UPG]|1[258]$",
            "type": "string",
          },
          "title": {
            "description": "The movie title (alphanumeric characters and spaces only)",
            "maxLength": 100,
            "minLength": 1,
            "pattern": "^[a-zA-Z0-9 ]*$",
            "type": "string",
          },
          "year": {
            "description": "The release year of the movie",
            "pattern": "^\\d{4}$",
            "type": "string",
          },
        },
        "required": [
          "id",
          "title",
          "year",
          "rating",
        ],
        "type": "object",
      },
      "NewMovie": {
        "properties": {
          "rating": {
            "description": "The rating of the movie",
            "enum": [
              "U",
              "PG",
              "12",
              "15",
              "18",
            ],
            "pattern": "^[UPG]|1[258]$",
            "type": "string",
          },
          "title": {
            "description": "The movie title (alphanumeric characters and spaces only)",
            "maxLength": 100,
            "minLength": 1,
            "pattern": "^[a-zA-Z0-9 ]*$",
            "type": "string",
          },
          "year": {
            "description": "The release year of the movie",
            "pattern": "^\\d{4}$",
            "type": "string",
          },
        },
        "required": [
          "title",
          "year",
          "rating",
        ],
        "type": "object",
      },
    },
  },
  "info": {
    "description": "An API for movies",
    "title": "Movie API Example",
    "version": "v1",
  },
  "openapi": "3.0.3",
  "paths": {
    "/v1/movies": {
      "get": {
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "items": {
                    "$ref": "#/components/schemas/Movie",
                  },
                  "type": "array",
                },
              },
            },
            "description": "Successful response",
          },
          "400": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/ErrorResponse",
                },
              },
            },
            "description": "Bad Request",
          },
          "404": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/ErrorResponse",
                },
              },
            },
            "description": "Not Found",
          },
        },
        "summary": "Get all movies",
        "x-amazon-apigateway-integration": {
          "credentials": "arn:aws:iam::account-id:role/apiRole",
          "httpMethod": "POST",
          "passthroughBehavior": "when_no_match",
          "responses": {
            "default": {
              "responseTemplates": {
                "application/json": "",
              },
              "statusCode": "200",
            },
          },
          "type": "aws_proxy",
          "uri": {
            "Fn::Sub": "arn:aws:apigateway:\${AWS::Region}:lambda:path/2015-03-31/functions/\${ListMoviesLambda.Arn}/invocations",
          },
        },
        "x-amazon-apigateway-request-validator": "validation",
      },
      "post": {
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/NewMovie",
              },
            },
          },
          "required": true,
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/Movie",
                },
              },
            },
            "description": "Successful response",
          },
          "400": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/ErrorResponse",
                },
              },
            },
            "description": "Bad Request",
          },
          "404": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/ErrorResponse",
                },
              },
            },
            "description": "Not Found",
          },
        },
        "summary": "Create a new movie",
        "x-amazon-apigateway-integration": {
          "credentials": "arn:aws:iam::account-id:role/apiRole",
          "httpMethod": "POST",
          "passthroughBehavior": "when_no_match",
          "responses": {
            "default": {
              "responseTemplates": {
                "application/json": "",
              },
              "statusCode": "200",
            },
          },
          "type": "aws_proxy",
          "uri": {
            "Fn::Sub": "arn:aws:apigateway:\${AWS::Region}:lambda:path/2015-03-31/functions/\${CreateMovieLambda.Arn}/invocations",
          },
        },
        "x-amazon-apigateway-request-validator": "validation",
      },
    },
    "/v1/movies/{id}": {
      "get": {
        "parameters": [
          {
            "description": "ID of the movie",
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
        ],
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/Movie",
                },
              },
            },
            "description": "Successful response",
          },
          "400": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/ErrorResponse",
                },
              },
            },
            "description": "Bad Request",
          },
          "404": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/ErrorResponse",
                },
              },
            },
            "description": "Not Found",
          },
        },
        "summary": "Get a movie by id",
        "x-amazon-apigateway-integration": {
          "credentials": "arn:aws:iam::account-id:role/apiRole",
          "httpMethod": "POST",
          "passthroughBehavior": "when_no_match",
          "responses": {
            "default": {
              "responseTemplates": {
                "application/json": "",
              },
              "statusCode": "200",
            },
          },
          "type": "aws_proxy",
          "uri": {
            "Fn::Sub": "arn:aws:apigateway:\${AWS::Region}:lambda:path/2015-03-31/functions/\${GetMovieByIdLambda.Arn}/invocations",
          },
        },
        "x-amazon-apigateway-request-validator": "validation",
      },
    },
  },
  "x-amazon-apigateway-request-validators": {
    "validation": {
      "validateRequestBody": true,
      "validateRequestParameters": true,
    },
  },
}
`;

exports[`get-api-json should return the redacted version of the json 1`] = `
{
  "components": {
    "schemas": {
      "ErrorResponse": {
        "properties": {
          "message": {
            "type": "string",
          },
        },
        "required": [
          "message",
        ],
        "type": "object",
      },
      "Movie": {
        "properties": {
          "id": {
            "description": "The movie ID (numeric characters only)",
            "pattern": "^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-4[0-9a-fA-F]{3}-[89aAbB][0-9a-fA-F]{3}-[0-9a-fA-F]{12}$",
            "type": "string",
          },
          "rating": {
            "description": "The rating of the movie",
            "enum": [
              "U",
              "PG",
              "12",
              "15",
              "18",
            ],
            "pattern": "^[UPG]|1[258]$",
            "type": "string",
          },
          "title": {
            "description": "The movie title (alphanumeric characters and spaces only)",
            "maxLength": 100,
            "minLength": 1,
            "pattern": "^[a-zA-Z0-9 ]*$",
            "type": "string",
          },
          "year": {
            "description": "The release year of the movie",
            "pattern": "^\\d{4}$",
            "type": "string",
          },
        },
        "required": [
          "id",
          "title",
          "year",
          "rating",
        ],
        "type": "object",
      },
      "NewMovie": {
        "properties": {
          "rating": {
            "description": "The rating of the movie",
            "enum": [
              "U",
              "PG",
              "12",
              "15",
              "18",
            ],
            "pattern": "^[UPG]|1[258]$",
            "type": "string",
          },
          "title": {
            "description": "The movie title (alphanumeric characters and spaces only)",
            "maxLength": 100,
            "minLength": 1,
            "pattern": "^[a-zA-Z0-9 ]*$",
            "type": "string",
          },
          "year": {
            "description": "The release year of the movie",
            "pattern": "^\\d{4}$",
            "type": "string",
          },
        },
        "required": [
          "title",
          "year",
          "rating",
        ],
        "type": "object",
      },
    },
  },
  "info": {
    "description": "An API for movies",
    "title": "Movie API Example",
    "version": "v1",
  },
  "openapi": "3.0.3",
  "paths": {
    "/v1/movies": {
      "get": {
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "items": {
                    "$ref": "#/components/schemas/Movie",
                  },
                  "type": "array",
                },
              },
            },
            "description": "Successful response",
          },
          "400": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/ErrorResponse",
                },
              },
            },
            "description": "Bad Request",
          },
          "404": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/ErrorResponse",
                },
              },
            },
            "description": "Not Found",
          },
        },
        "summary": "Get all movies",
        "x-amazon-apigateway-request-validator": "validation",
      },
      "post": {
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/NewMovie",
              },
            },
          },
          "required": true,
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/Movie",
                },
              },
            },
            "description": "Successful response",
          },
          "400": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/ErrorResponse",
                },
              },
            },
            "description": "Bad Request",
          },
          "404": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/ErrorResponse",
                },
              },
            },
            "description": "Not Found",
          },
        },
        "summary": "Create a new movie",
        "x-amazon-apigateway-request-validator": "validation",
      },
    },
    "/v1/movies/{id}": {
      "get": {
        "parameters": [
          {
            "description": "ID of the movie",
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
        ],
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/Movie",
                },
              },
            },
            "description": "Successful response",
          },
          "400": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/ErrorResponse",
                },
              },
            },
            "description": "Bad Request",
          },
          "404": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/ErrorResponse",
                },
              },
            },
            "description": "Not Found",
          },
        },
        "summary": "Get a movie by id",
        "x-amazon-apigateway-request-validator": "validation",
      },
    },
  },
  "x-amazon-apigateway-request-validators": {
    "validation": {
      "validateRequestBody": true,
      "validateRequestParameters": true,
    },
  },
}
`;
